service/history/ndc/hsm_state_replicator.go

```diff
@@ -183,10 +183,19 @@ func (r *HSMStateReplicatorImpl) syncHSMNode(
 		incomingNodePath := incomingNode.Path()
 		currentNode, err := currentHSM.Child(incomingNodePath)
 		if err != nil {
-			// 1. Already done history resend if needed before,
-			// and node creation today always associated with an event
-			// 2. Node deletion is not supported right now.
-			// Based on 1 and 2, node should always be found here.
+			// The node may not be found if:
+			// 1. The state machine was deleted (e.g. terminal state cleanup)
+			// 2. We're missing events that created this node
+			if errors.Is(err, hsm.ErrStateMachineNotFound) {
+				// In terminal state, nodes can be deleted
+				// Ignore the error and continue processing other nodes
+				r.logger.Debug("State machine not found - likely deleted in terminal state",
+					tag.WorkflowNamespaceID(mutableState.GetExecutionInfo().NamespaceId),
+					tag.WorkflowID(mutableState.GetExecutionInfo().WorkflowId),
+					tag.WorkflowRunID(mutableState.GetExecutionInfo().OriginalExecutionRunId),
+				)
+				return nil
```

Comment:
nit: I'd add a sanity check that version history in the mutable state is > the one in the request (same as the one on L265. or just return that info from compareVersionHistory), and return an error otherwise.

Reply:
Don't we already check this in compareVersionHistory?

Reply:
hmm not sure I follow. Error is returned from compareVersionHistory if last version history item of the (local) mutable state is < that in the request. The check I mentioned is for > (also not the same as the >= checked in compareVersionHistory)